google-api-client/src/main/java/com/google/api/client/googleapis/auth/oauth2/OAuth2Utils.java

```diff
@@ -53,20 +69,46 @@ static boolean headersContainValue(HttpHeaders headers, String headerName, Strin
     return false;
   }
 
-  static boolean runningOnComputeEngine(HttpTransport transport) {
-    try {
-      GenericUrl tokenUrl = new GenericUrl(METADATA_SERVER_URL);
-      HttpRequest request = transport.createRequestFactory().buildGetRequest(tokenUrl);
-      HttpResponse response = request.execute();
-      HttpHeaders headers = response.getHeaders();
-      if (headersContainValue(headers, "Metadata-Flavor", "Google")) {
-        return true;
+  static boolean runningOnComputeEngine(HttpTransport transport,
+      SystemEnvironmentProvider environment) {
+    // If the environment has requested that we do no GCE checks, return immediately.
+    if (Boolean.parseBoolean(environment.getEnv("NO_GCE_CHECK"))) {
+      return false;
+    }
+
+    GenericUrl tokenUrl = new GenericUrl(getMetadataServerUrl(environment));
+    for (int i = 1; i <= MAX_COMPUTE_PING_TRIES; ++i) {
+      try {
+        HttpRequest request = transport.createRequestFactory().buildGetRequest(tokenUrl);
+        request.setConnectTimeout(COMPUTE_PING_CONNECTION_TIMEOUT_MS);
+        HttpResponse response = request.execute();
+        try {
+          HttpHeaders headers = response.getHeaders();
+          return headersContainValue(headers, "Metadata-Flavor", "Google");
```

Comment:
Then please provide useful constant names; I don't have any. The ones I've thought of make the code unreadable, like `GOOGLE` and `MAGIC_VALUE` and `GOOGLE_FLAVOR`. I dislike the constants when they reduce readability and encourage bugs because they added a harmful layer of abstraction. When the new names are useful and nicely abstracted then it improves the readability and can reduce the likelihood of bugs.